### PR TITLE
NIP-30: Custom Emoji

### PR DESCRIPTION
**Pull request by alexgleason**
Adds a NIP for custom emoji, similar to Mastodon. It is [already implemented by the Mostr bridge](https://gitlab.com/soapbox-pub/mostr/-/merge_requests/56).

A real-world event can be found here: https://www.nostr.guru/e/5e1e0e524432e524f9584f296f72dda5559fc9fe6116a744baa43ea6769c7b65

### CODE REVIEW

**Comment by jb55** [comment]
LGTM

**Review by benthecarman — APPROVED** [review]

**Review by verbiricha — APPROVED** [review]

**Comment by alexgleason** [comment]
I've also implemented this [in Ditto](https://gitlab.com/soapbox-pub/ditto/-/commit/837a3ae1dda3457535bd83d5d27934f92f393423) now, so it's interoperating between two Nostr projects. MR has 3 approvals. Any other feedback? Merge?

**Comment by fiatjaf** [comment]
This should be a different event kind.

**Comment by fiatjaf** [comment]
Can we change?

**Comment by mikedilger** [comment]
I think a new event kind would just cause notes with custom emojis in them to drop entirely.

But IMHO every special thing inserted into a kind-1 note should IMHO ideally be inserted by a tag that specifies the offset to insert at, rather than infecting the content with a placeholder.  That includes ':emoji:', '#[1]', 'nostr:nevent1...', etc.  That way the content can still be sensibly rendered by clients that do not understand the tag, and there would be no need to parse the content.

**Comment by fiatjaf** [comment]
Nevermind, I thought you would define your emoji pack once and reuse everywhere, but since you have to specify the emoji URL every time it's good as it is.

**Comment by alexgleason** [comment]
I thought about that @mikedilger, but with NIP-27 there's nothing left that uses the `#[0]` scheme anymore. Also, a placeholder like `:nostrich:` is able to be typed by a human (with good autocomplete support) and displays nicely in unsupported clients.

**Comment by fiatjaf** [comment]
> But IMHO every special thing inserted into a kind-1 note should IMHO ideally be inserted by a tag that specifies the offset to insert at, rather than infecting the content with a placeholder. That includes ':emoji:', '#[1]', 'nostr:nevent1...', etc. That way the content can still be sensibly rendered by clients that do not understand the tag, and there would be no need to parse the content.

What if we had both?

**Comment by fiatjaf** [comment]
The emoji tag could be like `["emoji", "ablobcatrainbow", "https://gleasonator.com/emoji/blobcat/ablobcatrainbow.png", 7]`, 7 being the offset.

**Comment by mikedilger** [comment]
> I thought about that @mikedilger, but with NIP-27 there's nothing left that uses the `#[0]` scheme anymore. Also, a placeholder like `:nostrich:` is able to be typed by a human (with good autocomplete support) and displays nicely in unsupported clients.

Yes when composing you could type :nostrich: and have that replaced by a tag by the client before sending.  Similar to @alexgleason or #3e470a etc.

Maybe we can have a new textnote kind that does this.  Current text notes will always need to be parsed as there will always be older software around.

**Comment by staab** [comment]
Maybe we should have a convention of saying "last call" before merging something. Or maybe merging a PR just creates the urgency needed to actually have the discussion.

**Pull request merged**
